rfcs/007-bytecode-interpreter.md

```diff
@@ -0,0 +1,839 @@
+
+Because tables are used to represent arrays, Lua 5 uses a hybrid representation
+with an array part and a hash part. Once again, this is irrelevant for Nickel.
+
+Lua also has a peculiar representation of closures. There is a vast body of
+literature around compilation of closures to native code in functional
+languages, but the analyses involved can be expensive or too complex for a
+simple one-pass compiler. Instead, in Lua, outer variables are accessed
+indirectly through a slot called an _upvalue_, which originally points to the
+stack of the enclosing function owning the referenced value. When the stack
+frame is freed, the value is moved to the slot so that it can outlive the
+enclosing function call. Thanks to the additional indirection, this move is
+invisible to the closures themselves. This is combined with flat closures
+(access to more than one outer level causes the corresponding variable to be
+pulled in the enclosing function), and a list of open upvalues to ensure that
+there's only on such slot pointing to one given value at all time, even if this
+slot is shared by many closures.
+
+Compared to the traditional closure conversion of Haskell, it seems that the
+advantage is to avoid copying as much as possible (given Lua's "large" value
+representation). First, many closures can share the same slot, instead of each
+having their own copy of the captured value. Second, I suppose that if the
+enclosing function outlives the closure, no value is ever moved out from the
+slot, and thus no copying happens.
+
+On the other hand, this techniques looks useless if the values are represented
+on one word, as a pointer to the slot and the slot itself still need to be
+allocated, plus the machinery around that (list of open closures during
+compilation and at runtime to decide what should be moved when de-allocating a
+frame).
+
+#### Virtual machine
+
+The Lua virtual machine is register-based since Lua 5.0 (note that there's still
+a stack for activation records) while most bytecode VMs out there are usually
+stack-based, even if they might have a few registers. Lua developers argue that
+a register-based VM saves many `push` and `pop` operations, which both avoids a
+non-trivial amount of copying (especially with the "large" tagged union
+representation of values) and reduces the number of instructions (albeit the
+instructions are larger, because they often need additional operands to indicate
+which registers they operate on). Registers also reduce the number of
+instructions needed: the Lua VM only has 35. The Lua VM has 256 registers
+represented as a packed array at runtime, guaranteeing fast access.
+
+#### References
+
+- [The Implementation of LUA 5](https://www.lua.org/doc/jucs05.pdf)
+
+### OCaml
+
+#### Memory representation
+
+The following notes on the memory representation applies to the native code
+backend's representation. I'm not sure how closures are represented in the Zinc
+Abstract Machine.
```

Comment:
I'm not sure it ought to be called Zinc anymore, but anyway: I'm pretty sure that the representation of values (including closures) is the same in native and bytecode. It must be so at least to some degree for the sake of the FFI, where Ocaml values can be manipulated.